src/inmanta/server/services/compilerservice.py

```diff
@@ -763,21 +781,57 @@ async def notify(listener: CompileStateListener) -> None:
             except Exception:
                 logging.exception("CompileStateListener failed")
 
-        await asyncio.gather(*[notify(listener) for listener in self.listeners])
-        await compile.update_fields(handled=True)
-        LOGGER.log(const.LOG_LEVEL_TRACE, "listeners notified for %s", compile.id)
+        blockers = asyncio.gather(*[notify(listener) for listener in self.blocking_listeners])
+
+        async def _sub_notify() -> None:
+            await asyncio.gather(*[notify(listener) for listener in self.async_listeners])
+            await blockers
+            await compile.update_fields(handled=True)
+            LOGGER.log(const.LOG_LEVEL_TRACE, "listeners notified for %s", compile.id)
+
+        self.add_background_task(_sub_notify())
+        await blockers
 
     async def _recover(self) -> None:
         """Restart runs after server restart"""
+        # patch up in-memory state
         async with self._queue_count_cache_lock:
             self._queue_count_cache = await data.Compile.get_total_length_of_all_compile_queues(exclude_started_compiles=False)
-        # one run per env max to get started
-        runs = await data.Compile.get_next_run_all()
-        for run in runs:
-            await self._queue(run)
-        unhandled = await data.Compile.get_unhandled_compiles()
-        for u in unhandled:
-            self.add_background_task(self._notify_listeners(u))
+
+        # we run the recovery and start in the background to prevent deadlocks with sync handlers
+        async def sub_recovery() -> None:
+            unhandled = await data.Compile.get_unhandled_compiles()
+            for u in unhandled:
+                await self._notify_listeners(u)
+            # one run per env max to get started
+            runs = await data.Compile.get_next_run_all()
+
+            # prevent races by holding global lock
+            async with self._global_lock:
```

Comment:
Even with this lock I think there might be a race with `self._queue`: that method takes a compile as an argument. After acquiring the lock it does not check whether that compile may have already run.
1. server is started, `sub_recovery` runs as background task
2. `request_compile` is called for compile `c1`
3. `request_recompile` inserts compile in db and calls `self._queue(c1)`
4. `sub_recovery` wins race for global lock: discovers `c1` and starts running it
5. `c1` is a very fast compile and finishes almost immediately
6. `sub_recovery` has a bunch of other compiles to start because there are multiple environments. Only after that does it release the lock
7. event loop passes control back to `self._queue`, which acquires the lock. It sees that no compile is running for this environment, so it runs its argument `c1`, even though it has already run.

Reply:
This is probably extremely rare, but since it would be easy to address I think we should.

Reply:
Same in the other direction I think.

Reply:
From Slack (for future reference):

Ah, but perhaps the opposite is also possible?
1. recover fetches waiting compiles (`Compile.get_next_run_all()`) -> there are none because we're starting clean
2. compile is requested: `request_recompile` calls `self._queue(c1)` and wins the race -> returns without doing anything because `not self.fully_ready`
3. recovery acquires the lock, sets `fully_ready = True`, starts 0 compiles.
4. nothing happens. Compiles remain in the queue until a new one is requested, and only then do we start consuming again.